ports/raspberrypi/common-hal/wifi/Radio.c

```diff
@@ -0,0 +1,311 @@
+}
+
+mp_obj_t common_hal_wifi_radio_get_ipv4_subnet_ap(wifi_radio_obj_t *self) {
+    if (!netif_is_up(NETIF_AP)) {
+        return mp_const_none;
+    }
+    return common_hal_ipaddress_new_ipv4address(NETIF_AP->netmask.addr);
+}
+
+mp_obj_t common_hal_wifi_radio_get_ipv4_address(wifi_radio_obj_t *self) {
+    if (!netif_is_up(NETIF_STA)) {
+        return mp_const_none;
+    }
+    return common_hal_ipaddress_new_ipv4address(NETIF_STA->ip_addr.addr);
+}
+
+mp_obj_t common_hal_wifi_radio_get_ipv4_address_ap(wifi_radio_obj_t *self) {
+    if (!netif_is_up(NETIF_AP)) {
+        return mp_const_none;
+    }
+    return common_hal_ipaddress_new_ipv4address(NETIF_AP->ip_addr.addr);
+}
+
+mp_obj_t common_hal_wifi_radio_get_ipv4_dns(wifi_radio_obj_t *self) {
+    uint32_t addr = dns_getserver(0)->addr;
+    if (!netif_is_up(NETIF_STA) || addr == 0) {
+        return mp_const_none;
+    }
+    return common_hal_ipaddress_new_ipv4address(addr);
+}
+
+void common_hal_wifi_radio_set_ipv4_dns(wifi_radio_obj_t *self, mp_obj_t ipv4_dns_addr) {
+    mp_raise_NotImplementedError(NULL);
+}
+
+void common_hal_wifi_radio_start_dhcp_client(wifi_radio_obj_t *self) {
+}
+
+void common_hal_wifi_radio_stop_dhcp_client(wifi_radio_obj_t *self) {
+    mp_raise_NotImplementedError(NULL);
+}
+
+void common_hal_wifi_radio_set_ipv4_address(wifi_radio_obj_t *self, mp_obj_t ipv4, mp_obj_t netmask, mp_obj_t gateway, mp_obj_t ipv4_dns) {
+    mp_raise_NotImplementedError(NULL);
+}
+
+volatile bool ping_received;
+
+static u8_t
+ping_recv(void *arg, struct raw_pcb *pcb, struct pbuf *p, const ip_addr_t *addr) {
+    struct icmp_echo_hdr *iecho;
+    LWIP_ASSERT("p != NULL", p != NULL);
+
+    if ((p->tot_len >= (PBUF_IP_HLEN + sizeof(struct icmp_echo_hdr))) &&
+        pbuf_remove_header(p, PBUF_IP_HLEN) == 0) {
+        iecho = (struct icmp_echo_hdr *)p->payload;
+
+        if ((iecho->id == PING_ID) && (iecho->seqno == lwip_htons(ping_seq_num))) {
+            LWIP_DEBUGF(PING_DEBUG, ("ping: recv "));
+            ip_addr_debug_print(PING_DEBUG, addr);
+            LWIP_DEBUGF(PING_DEBUG, (" %"U32_F " ms\n", (sys_now() - ping_time)));
+
+            /* do some ping result processing */
+            ping_received = true;
+            pbuf_free(p);
+            return 1; /* eat the packet */
+        }
+        /* not eaten, restore original packet */
+        pbuf_add_header(p, PBUF_IP_HLEN);
+    }
+    return 0; /* don't eat the packet */
+}
+
+mp_int_t common_hal_wifi_radio_ping(wifi_radio_obj_t *self, mp_obj_t ip_address, mp_float_t timeout) {
+    ip_addr_t ping_addr;
+    ipaddress_ipaddress_to_lwip(ip_address, &ping_addr);
+
+    struct raw_pcb *ping_pcb;
+    MICROPY_PY_LWIP_ENTER
+        ping_pcb = raw_new(IP_PROTO_ICMP);
+    if (!ping_pcb) {
+        MICROPY_PY_LWIP_EXIT
+        return -1;
+    }
+    raw_recv(ping_pcb, ping_recv, NULL);
+    raw_bind(ping_pcb, IP_ADDR_ANY);
+    MICROPY_PY_LWIP_EXIT
+
+        ping_received = false;
```

Comment:
Fix indentations? You could probably add a semicolon after the `_ENTER` and `_EXIT`. I prefer these kinds of macros to be used with semicolons.

(This is a nit.)

Reply:
This code is copied pretty much verbatim from micropython. I don't know why they did it like that. The indentation is forced by uncrustify, I can't change it.